usecases/config/runtimeconfig.go

```
@@ -13,65 +13,123 @@ package config
+ 1. Everytime new config is supported, this function gets verbose as we have update for every struct fields in WeaviateRuntimeConfig
+ 2. The much bigger one is, what if consumer added a struct field, but failed to **update** this function?. This was a serious concern for me, more work for
+    consumers.
 
-	dec := yaml.NewDecoder(bytes.NewReader(buf))
-	dec.SetStrict(true)
+With this reflection method, we avoided that extra step from the consumer. This reflection approach is "logically" same as above implementation.
+See "runtimeconfig_test.go" for more examples.
+*/
+func updateRuntimeConfig(source, parsed reflect.Value) {
+	// Basically we do following
+	//
+	// 1. Loop through all the `source` fields
+	// 2. Check if any of those fields exists in `parsed` (non-nil)
+	// 3. If parsed config doesn't contain the field from `source`, We reset source's field.
+	//    so that it's default value takes preference.
+	// 4. If parsed config does contain the field from `source`, We update the value via `SetValue`.
 
-	// Am empty runtime yaml file is still a valid file. So treating io.EOF as
-	// non-error case returning default values of conf.
-	if err := dec.Decode(&conf); err != nil && !errors.Is(err, io.EOF) {
-		return nil, err
+	for i := range source.NumField() {
+		sf := source.Field(i)
+		pf := parsed.Field(i)
+
+		if pf.IsNil() {
+			method := sf.MethodByName("Reset")
+			if method.IsValid() && method.Type().NumIn() == 0 {
+				method.Call(nil)
+			}
+			continue
+		}
+
+		// NOTE(kavi):
+		// Q: why not type cast it into *DynamicValue[T] and call the method instead?
+		// A: Given DynamicValue[T] is generic type, you have to know the concrete type before doing type cast. Which is not possible here. ¯\_(ツ)_/¯
+		gmethod := pf.MethodByName("Get")
+		smethod := sf.MethodByName("SetValue")
+
+		if !gmethod.IsValid() || gmethod.Type().NumIn() != 0 || gmethod.Type().NumOut() != 1 {
+			panic("Get() has non-zero arguments and more than one return values")
```

**suggestion:** could we avoid panics and instead return an error ? the panic could happen at config time, but lets imagine there is a running instance and the config file got changed with typos in this case i'd say instead we error and avoid disturbing running instance by panic

---

But these panics are never "developer" or "operator" mistakes. This panic happens, if `Get()` or `SetValue()`  method signature changes and there no way these method can be called. 

Even if we return as error, to take your use case (typo in configs), nothing an operator can do to fix it. Hence the panic (the state we should never reach).

I added tests to lock this behavior. To catch such things. Would have been nicer if compiler catch these. But yea it's Go 🤷

---

hmm, wdyt if we use interfaces to simplify and avoid panics 

e.g. 
```go
// DynamicConfigValue represents a dynamic configuration value
type DynamicConfigValue[T any] interface {
	Reset()
	Get() T
	SetValue(T)
}

func UpdateRuntimeConfig(source, parsed *WeaviateRuntimeConfig) error {
	if source == nil || parsed == nil {
		return fmt.Errorf("source and parsed cannot be nil")
	}

	sourceVal := reflect.ValueOf(source).Elem()
	parsedVal := reflect.ValueOf(parsed).Elem()

	for i := 0; i < sourceVal.NumField(); i++ {
		sourceField := sourceVal.Field(i)
		parsedField := parsedVal.Field(i)

		if !sourceField.IsValid() || !parsedField.IsValid() {
			continue
		}

		sourceIface := sourceField.Interface()
		parsedIface := parsedField.Interface()

		// If parsed is nil, try to reset the source if it implements DynamicConfigValue
		if parsedField.IsNil() {
			if src, ok := sourceIface.(interface{ Reset() }); ok {
				src.Reset()
			}
			continue
		}

		switch s := sourceIface.(type) {
		case DynamicConfigValue[int]:
			if p, ok := parsedIface.(DynamicConfigValue[int]); ok {
				s.SetValue(p.Get())
			}
		case DynamicConfigValue[bool]:
			if p, ok := parsedIface.(DynamicConfigValue[bool]); ok {
				s.SetValue(p.Get())
			}
		... rest of the supported types 
		}
	}

	return nil
}

```

---

Even in your above we need panics in two places. 
1. what if `Reset()` interface type case fails? (just `continuing` is silent error, operators won't see config changes getting reflected. Worse)
2. what goes in the `default` case of your switch? (doesn't recognize any type).

I tweaked and covered those cases. Now at least we call the methods directly (compiler complains if those signature changes). So nice gain. Thanks for the suggestion :)